chromium_src/ios/web/webui/url_data_source_ios.mm

```diff
@@ -0,0 +1,154 @@
+#include "base/strings/utf_string_conversions.h"
+#include "ios/components/webui/web_ui_url_constants.h"
+#include "services/network/public/mojom/content_security_policy.mojom.h"
+
+namespace {
+
+// A chrome-untrusted data source's name starts with chrome-untrusted://.
+bool IsChromeUntrustedDataSource(const web::URLDataSourceIOS* source) {
+  static const base::NoDestructor<std::string> kChromeUntrustedSourceNamePrefix(
+      base::StrCat({kChromeUIUntrustedScheme, url::kStandardSchemeSeparator}));
+
+  return base::StartsWith(source->GetSource(),
+                          *kChromeUntrustedSourceNamePrefix,
+                          base::CompareCase::SENSITIVE);
+}
+
+}  // namespace
+
+namespace web {
+bool URLDataSourceIOS::ShouldAddContentSecurityPolicy() const {
+  return true;
+}
+
+bool URLDataSourceIOS::ShouldServiceRequest(const GURL& url) const {
+  return URLDataSourceIOS::ShouldServiceRequest_ChromiumImpl(url);
+}
+
+std::string URLDataSourceIOS::GetContentSecurityPolicyObjectSrc() const {
+  if (ShouldAddContentSecurityPolicy()) {
+    std::string csp_header;
+
+    const network::mojom::CSPDirectiveName kAllDirectives[] = {
+        network::mojom::CSPDirectiveName::BaseURI,
+        network::mojom::CSPDirectiveName::ChildSrc,
+        network::mojom::CSPDirectiveName::ConnectSrc,
+        network::mojom::CSPDirectiveName::DefaultSrc,
+        network::mojom::CSPDirectiveName::FencedFrameSrc,
+        network::mojom::CSPDirectiveName::FormAction,
+        network::mojom::CSPDirectiveName::FontSrc,
+        network::mojom::CSPDirectiveName::ImgSrc,
+        network::mojom::CSPDirectiveName::MediaSrc,
+        network::mojom::CSPDirectiveName::ObjectSrc,
+        network::mojom::CSPDirectiveName::RequireTrustedTypesFor,
+        network::mojom::CSPDirectiveName::ScriptSrc,
+        network::mojom::CSPDirectiveName::StyleSrc,
+        network::mojom::CSPDirectiveName::TrustedTypes,
+        network::mojom::CSPDirectiveName::WorkerSrc};
+
+    for (auto& directive : kAllDirectives) {
+      csp_header.append(GetContentSecurityPolicy(directive));
+    }
+
+    // TODO(crbug.com/40118579): Both CSP frame ancestors and XFO headers may be
+    // added to the response but frame ancestors would take precedence. In the
+    // future, XFO will be removed so when that happens remove the check and
+    // always add frame ancestors.
+    if (ShouldDenyXFrameOptions()) {
+      csp_header.append(GetContentSecurityPolicy(
+          network::mojom::CSPDirectiveName::FrameAncestors));
+    }
+
+    return csp_header;
+  }
+
+  return URLDataSourceIOS::GetContentSecurityPolicyObjectSrc_ChromiumImpl();
+}
+
+std::string URLDataSourceIOS::GetContentSecurityPolicyFrameSrc() const {
+  std::string frame_src =
+      GetContentSecurityPolicy(network::mojom::CSPDirectiveName::FrameSrc);
+  if (!frame_src.empty()) {
+    return frame_src;
+  }
+
+  return "frame-src 'none';";
+}
+
+std::string URLDataSourceIOS::GetContentSecurityPolicy(
```

Comment:
it looks like a lot of this stuff is copied from `URLDataSource` which we should not be doing because it's very fragile.